File: kth_smallest_element_in_BST.cpp
//using priority queue 
//Time Complexity-O(n)-->'n' is number of nodes
//Space Complexity-O(k)-->'k' is queue size
//Did the code execute on Leetcode? Yes

class Solution {
public:
    void small(TreeNode* root, int &k,priority_queue<int,vector<int>>&m)
    {
        if(root==NULL)
        {
            return;
        }
        if(k!=0)
        {
            m.push(root->val);
            k--;
        }
        else
        {
            if(root->val<m.top())
            {
                m.pop();
                m.push(root->val);
            }
        }
        if(root->left==NULL && root->right==NULL)
        {
            return;
        }
        small(root->left,k,m);
        small(root->right,k,m);
        
    }
    
    int kthSmallest(TreeNode* root, int k) {
        priority_queue<int,vector<int>>m;
        small(root,k,m);
        return m.top();
    }
};

File: kth smallest element inbst.java
//Timecomplexity:- O(n);
//spacecomplexity:-O(n)
//did it run on leetcode:- Yes.
//Your code with explanation:- in order traversal of bst gives ascending order storing them in min heap and removing root until k1 equals k.




class Solution {
    ArrayList<Integer> cache=new ArrayList<>();
    public int kthSmallest(TreeNode root, int k) {
       
        PriorityQueue<Integer> minheap=new PriorityQueue<>();
            inorder(root);
        int k1=1;
        for(int m=0;m<cache.size();m++){
            minheap.add(cache.get(m));
        }
        while(k1<k){
            minheap.poll();
            k1++;
        }
        return minheap.peek();
    }
    private void inorder(TreeNode root){
        if(root==null){
            return;
        }
        inorder(root.left);
        cache.add(root.val);
        inorder(root.right);
    }
}